src/group_theory/group_ring.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2021 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amelia Livingston
+-/
+
+import algebra.monoid_algebra.basic
+import tactic.basic
+
+/-!
+# Group rings
+
+This file defines the group ring `ℤ[G]` of a group `G`.
+
+This is the free abelian group on the elements of `G`, with multiplication induced by
+multiplication in `G.` A `G`-module `M` is also a `ℤ[G]`-module.
+
+We develop an API allowing us to show `ℤ[Gⁿ]` is a free `ℤ[G]`-module for `n ≥ 1`. This
+will be used to construct a projective resolution of the trivial `ℤ[G]`-module `ℤ`.
+
+## Implementation notes
+
+Although `group_ring G` is just `monoid_algebra ℤ G`, we make the definition `group_ring`
+so Lean finds the right instances for this setting, and so we can separate material relevant
+to group cohomology into the namespace `group_ring`.
```

Comment:
To what extent can all of this be done for an arbitrary (comm) ring? There's no need to specialize to `ℤ` already, is there?

Reply:
This is more subtle than you might think. Here are two ways of defining the adeles: (1) define the adeles of Q as a concrete type in Type with no inputs (like the rationals and the reals; after all, we have the p-adics as a concrete type as well). Now define the adeles of K as just K tensor (adeles of Q). (2) Define the adeles of a number field as the restricted product over places in the usual way. Which is "better"? I don't really know. Mathematically they're the same of course. Here we can define the concrete group ring over Z as an object (depending on G, but everything depends on G here) and then for R-modules we can just carry the R around independently, or we can just tensor up to R. Alternatively we can build everything over R from the start. The disadvantage of building everything over R and not doing the Z case is that when we're faced with an add_comm_group we have to promote it to a Z-module. Is that the sort of thing which causes problems nowadays? 

Here is a practical response. Johan is suggesting another approach, where we avoid making the concrete instance with Z and just work with general R. One could then develop the entire theory using both approaches and find out which one works best. However my understanding is that Amelia has already developed a huge amount of the theory using this approach and has shown that it works. On the other hand I don't know how much she's thought about the cohomology of R-modules. One way of approaching the cohomology of R-modules (i.e. of proving that the cohomology of an R-module is an R-module) is by saying "take the cohomology of the underlying Z-module and then note that everything is functorial so the R-module action on M induces an R-module action on the cohomology". Another approach is to do everything over R from the offset and hope that when you specialise to R=Z there are no weird diamonds. A third approach is to do it the way Amelia has done and then explicitly put R-actions on the cocycles, the coboundaries etc, and this sounds feasible to me (although if one has R-actions on abelian groups Z and B does one get an easy action of R on the abelian group quotient Z/B? That might be the only issue I'd worry about). 

All of these ways are ways to approach group cohomology and unless someone tries all three I think it would be very difficult to figure out which approach is best.

Reply:
> Is that the sort of thing which causes problems nowadays?

I'm inclined to say that it doesn't cause any problems at all these days.

Reply:
I agree that the theory should be built up over a general ring from the start. I don't think it's that much of an issue going between additive groups and `Z`-modules, and even if it is, it certainly cannot be more annoying than wrangling a bunch of tensor products.